src/types/impl-trait.md

```diff
@@ -111,10 +133,193 @@ fn foo() -> impl Trait {
 doesn't allow the caller to determine the return type.
 Instead, the function chooses the return type, but only promises that it will implement `Trait`.
 
+
+## Type Alias Impl Trait
+
+In contrast to `impl Trait`s in function return types, type aliases can be used in more places than just return types and the same type alias can be used multiple times.
+
+```rust,ignore
+fn foo() -> impl Trait {
+    value_of_type_that_implements_Trait
+}
+
+// The above function can be changed to the
+// following without a change in behaviour for callers.
+
+type Foo = impl Trait;
+fn foo() -> Foo {
+    value_of_type_that_implements_Trait
+}
+```
+
+When such a type alias is used as the return type of multiple functions, all functions must use the same hidden type.
+This is similar to how all code paths in a function with a return-position-impl-trait must return the same type:
+
+```rust,ignore
+fn foo() -> impl Debug {
+    if true {
+        return 42;
+    }
+    "42" // ERROR: another return site returns an `i32`
+}
+
+type Bar = impl Debug;
+fn bar() -> Bar {
+    42
+}
+fn boo() -> Bar {
+    "42" // ERROR: another function uses `i32` for `Bar`
+}
+```
+
+### Argument position
+
+Without knowing the hidden type, we can still use the opaque type and use its trait bounds. In this case we can use the `Debug` trait to render it:
+
+```rust,ignore
+fn bop(x: Bar) {
+    println!("{x:?}");
+}
+```
+
+Note that this is very different from using `impl Trait` in argument position, as there is no anonymous generic parameter introduced.
+
+Defining a hidden type works in both directions, not just assigning a hidden type value to the opaque type, but also reading an opaque type into a hidden type value:
+
+```rust,ignore
+fn bup(x: Bar) {
+    let x: i32 = x;
+}
+```
+
+This does not "reveal" the hidden type. It binds an explicitly known `i32` type as the hidden type of `Bar` and
+will error if any of the hidden type defining sites use a different type.
+
+As a last usage, you can avoid defining any hidden types and just use the type-alias-impl-trait by just forwarding it elsewhere:
+
+```rust,ignore
+fn burp(x: Bar) -> Bar {
+    x
+}
+```
+
+### Binding types
+
+You can also use type-alias-impl-trait for the type
+of local variables, constants, statics, ...
```

Comment:
Is `...` here to be completed?

Also, when possible it is preferred to link to concepts (even if they seem painfully obvious to you).

```suggestion
of [local variables], [constants], [statics], ...

[local variables]: ../variables.md
[constants]: ../items/constant-items.md
[statics]: ../items/static-items.md
```